src/main/java/com/dynatrace/openkit/protocol/Beacon.java

```diff
@@ -698,32 +724,61 @@ private String createImmutableBasicBeaconData() {
         addKeyValuePair(basicBeaconBuilder, BEACON_KEY_APPLICATION_ID, configuration.getApplicationID());
         addKeyValuePair(basicBeaconBuilder, BEACON_KEY_APPLICATION_NAME, configuration.getApplicationName());
-        if (configuration.getApplicationVersion() != null) {
-            addKeyValuePair(basicBeaconBuilder, BEACON_KEY_APPLICATION_VERSION, configuration.getApplicationVersion());
-        }
+        addKeyValuePairIfNotNull(basicBeaconBuilder, BEACON_KEY_APPLICATION_VERSION, configuration.getApplicationVersion());
         addKeyValuePair(basicBeaconBuilder, BEACON_KEY_PLATFORM_TYPE, ProtocolConstants.PLATFORM_TYPE_OPENKIT);
         addKeyValuePair(basicBeaconBuilder, BEACON_KEY_AGENT_TECHNOLOGY_TYPE, ProtocolConstants.AGENT_TECHNOLOGY_TYPE);
 
         // device/visitor ID, session number and IP address
-        addKeyValuePair(basicBeaconBuilder, BEACON_KEY_VISITOR_ID, configuration.getDeviceID());
-        addKeyValuePair(basicBeaconBuilder, BEACON_KEY_SESSION_NUMBER, sessionNumber);
+        addKeyValuePair(basicBeaconBuilder, BEACON_KEY_VISITOR_ID, getVisitorID(configuration));
+        addKeyValuePair(basicBeaconBuilder, BEACON_KEY_SESSION_NUMBER, getSessionID(sessionNumber));
         addKeyValuePair(basicBeaconBuilder, BEACON_KEY_CLIENT_IP_ADDRESS, clientIPAddress);
 
         // platform information
-        if (configuration.getDevice().getOperatingSystem() != null) {
-            addKeyValuePair(basicBeaconBuilder, BEACON_KEY_DEVICE_OS, configuration.getDevice().getOperatingSystem());
-        }
-        if (configuration.getDevice().getManufacturer() != null) {
-            addKeyValuePair(basicBeaconBuilder, BEACON_KEY_DEVICE_MANUFACTURER, configuration.getDevice()
-                                                                                             .getManufacturer());
-        }
-        if (configuration.getDevice().getModelID() != null) {
-            addKeyValuePair(basicBeaconBuilder, BEACON_KEY_DEVICE_MODEL, configuration.getDevice().getModelID());
-        }
+        addKeyValuePairIfNotNull(basicBeaconBuilder, BEACON_KEY_DEVICE_OS, configuration.getDevice().getOperatingSystem());
+        addKeyValuePairIfNotNull(basicBeaconBuilder, BEACON_KEY_DEVICE_MANUFACTURER, configuration.getDevice().getManufacturer());
+        addKeyValuePairIfNotNull(basicBeaconBuilder, BEACON_KEY_DEVICE_MODEL, configuration.getDevice().getModelID());
 
         return basicBeaconBuilder.toString();
     }
 
+    /**
+     * Get a visitor ID for the current data collection level
+     *
+     * in case of level 2 (USER_BEHAVIOR) the value from the configuration is used
+     * in case of level 1 (PERFORMANCE) or 0 (OFF) a random number in the positive Long range is used
+     *
+     * @param configuration configuration object
+     * @return
+     */
+    public long getVisitorID(Configuration configuration){
+
+        if(configuration != null && beaconConfiguration.get() != null)
```

Comment:
in theory, there could be a race condition between `beaconConfiguration.get() != null` and the next access to `beaconConfiguration`, since a second thread could set it to `null` in the mean time.

Luckily we do not allow null values when setting the `BeaconConfiguration`.

If you would like to be on the safe side, please use the following pattern
``` java
BeaconConfiguration configuration = beaconConfiguration.get(); // obtain the reference
if (configuration != null) {
    // for sure it's not null here
}